File: bindings/aitp-py/tests/test_delegation.py
"""Delegation (RFC-AITP-0006) — Python SDK.

Covers the v0.2 single-hop round-trip and the strict-by-default posture:
the default `verify_delegation` rejects any draft RFC-AITP-0011 multi-hop
chain, and the opt-in `verify_delegation_multihop` (only present
when the binding is built with `--features multihop-delegation`)
is the sole way to accept one.

v0.2 wire shape: TCTs, grant vouchers, and delegation tokens are all opaque
**compact JWS strings**. B delegates from the *grant voucher* it received
alongside its TCT in the handshake commit — not from the TCT itself.

Run with `maturin develop` then `pytest` from
`bindings/aitp-py/`.
"""

import base64
import json

import pytest

import aitp

PING_CAP = "demo.write"


def _b64url_decode(seg: str) -> bytes:
    return base64.urlsafe_b64decode(seg + "=" * (-len(seg) % 4))


def _b64url_encode(raw: bytes) -> str:
    return base64.urlsafe_b64encode(raw).rstrip(b"=").decode("ascii")


def _build_delegation_token():
    """A issues a TCT (+ grant voucher) to B via an in-process handshake; B
    mints a single-hop delegation token (compact JWS) authorizing C. Returns
    (a, b, c, delegation_token)."""
    a = aitp.AitpAgent.generate()
    b = aitp.AitpAgent.generate()
    c = aitp.AitpAgent.generate()

    a_manifest = a.build_manifest(
        display_name="A",
        handshake_endpoint="http://localhost:8100/aitp/handshake/",
        offered_caps=[PING_CAP],
    )
    b.build_manifest(
        display_name="B",
        handshake_endpoint="http://localhost:8101/aitp/handshake/",
        offered_caps=[PING_CAP],
    )
    c.build_manifest(
        display_name="C",
        handshake_endpoint="http://localhost:8102/aitp/handshake/",
        offered_caps=[PING_CAP],
    )

    # B handshakes against A → B holds A's TCT + grant voucher for PING_CAP.
    bsess = b.new_session()
    arsess = a.new_responder()
    hello = bsess.build_hello(a_manifest, [PING_CAP])
    hello_ack, sid = arsess.process_hello(hello)
    commit = bsess.process_hello_ack(hello_ack, sid)
    commit_ack, _ = arsess.process_commit(commit)
    b_completed = json.loads(bsess.complete(commit_ack))
    b_voucher = b_completed["grant_voucher"]
    assert b_voucher is not None

    # B mints a delegation token rooted in A's voucher, binding C, scoped to
    # PING_CAP. C's key binding is derived from its AID — no pubkey arg.
    delegation_token = b.build_delegation(b_voucher, c.aid, [PING_CAP])
    return a, b, c, delegation_token


def test_delegation_round_trip_strict():
    """RFC-AITP-0006 single-hop: A verifies B's delegation under the strict
    default verifier and mints a fresh TCT for C, which C can verify."""
    a, b, c, delegation_token = _build_delegation_token()

    verified = aitp.verify_delegation(delegation_token, a.aid)
    assert verified.delegator == a.aid
    assert verified.delegatee == c.aid
    assert verified.issued_by == b.aid
    assert verified.grants == [PING_CAP]

    # A mints a fresh TCT for C; completion shape is {"tct", "grant_voucher"}.
    fresh = json.loads(a.issue_tct_for_delegatee(verified))
    ident = c.verify_tct(fresh["tct"], PING_CAP, c.aid)
    assert ident.peer_aid == a.aid
    assert PING_CAP in ident.grants


def test_verify_delegation_rejects_wrong_verifier():
    a, _b, _c, delegation_token = _build_delegation_token()
    other = aitp.AitpAgent.generate()
    with pytest.raises(Exception):
        aitp.verify_delegation(delegation_token, other.aid)


def _inject_multihop_chain(delegation_token):
    """Take a valid single-hop compact JWS and inject a non-empty `chain`
    claim into its (unsigned) payload, re-encoding the segment.

    The signature no longer matches, but the strict-vs-multihop hop gate
    fires on the structural `chain` presence *before* any signature work
    (RFC-AITP-0006 §4.4), so this is enough to exercise that gate. We reuse
    the token itself as a bogus chain entry to keep `chain` a non-empty list.
    """
    header_seg, payload_seg, sig_seg = delegation_token.split(".")
    payload = json.loads(_b64url_decode(payload_seg))
    payload["chain"] = [delegation_token]
    new_payload_seg = _b64url_encode(
        json.dumps(payload, separators=(",", ":")).encode("utf-8")
    )
    return f"{header_seg}.{new_payload_seg}.{sig_seg}"


def test_strict_verify_rejects_multihop_chain():
    """The default `verify_delegation` is strict v0.2: any non-empty chain is
    rejected with DELEGATION_MULTIHOP_NOT_SUPPORTED before any per-hop work
    (RFC-AITP-0006 §4.4)."""
    a, _b, _c, delegation_token = _build_delegation_token()
    tampered = _inject_multihop_chain(delegation_token)

    with pytest.raises(Exception) as exc:
        aitp.verify_delegation(tampered, a.aid)
    assert "multi-hop delegation is not supported" in str(exc.value)


def test_multihop_opts_past_the_hop_gate():
    """The opt-in verifier (built under `multihop-delegation`)
    must get PAST the hop gate the strict path rejects at — proven by it
    failing with a *different* error (structure/signature) rather than
    MULTIHOP_NOT_SUPPORTED."""
    if not hasattr(aitp, "verify_delegation_multihop"):
        pytest.skip("binding built without --features multihop-delegation")

    a, _b, _c, delegation_token = _build_delegation_token()
    tampered = _inject_multihop_chain(delegation_token)

    with pytest.raises(Exception) as exc:
        aitp.verify_delegation_multihop(tampered, a.aid, 3)
    assert "multi-hop delegation is not supported" not in str(exc.value)

File: bindings/aitp-node/tests/test_delegation.mjs
// Delegation (RFC-AITP-0006, voucher-based v0.2) round-trip — Node SDK.
//
// Three agents: A (issuer/verifier), B (delegator), C (delegatee).
//   1. A issues a TCT + grant voucher to B for ["demo.write"] via the
//      four-message handshake.
//   2. B mints a delegation token from its grant voucher, binding C's AID.
//   3. A verifies the delegation, then mints a fresh TCT for C.
//   4. C verifies the fresh TCT under the presented-TCT model.

import test from 'node:test';
import assert from 'node:assert/strict';

import {
  AitpAgent,
  verifyDelegation,
  verifyDelegationMultihop,
} from '../index.js';
import {
  withClaims,
  decodeJwsPayload,
  tamperJwsSignature,
} from './_jws.mjs';

const HAS_MULTIHOP =
  typeof verifyDelegationMultihop === 'function';

// Take a valid single-hop delegation token and forge a non-empty `chain`
// claim into its (unverified) payload. The strict vs. multi-hop gate
// fires on the decoded `chain` *before* the signature check, so the stale
// signature is irrelevant to the gate — enough to exercise it.
function injectMultihopChain(delegationToken) {
  return withClaims(delegationToken, { chain: [delegationToken] });
}

// Run the handshake and return B's grant voucher (the delegation root) for
// the requested grants.
function buildVoucher() {
  const { agent: a, manifest: aManifest } = buildPeer('A', 8301, ['demo.write']);
  const { agent: b } = buildPeer('B', 8302, ['demo.echo']);
  const { agent: c } = buildPeer('C', 8303, ['demo.read']);

  const bCompleted = fullHandshake(b, a, aManifest, ['demo.write']);
  const delegationToken = b.buildDelegation(
    bCompleted.grantVoucher,
    c.aid,
    ['demo.write'],
  );
  return { a, delegationToken };
}

function buildPeer(name, port, offers) {
  const agent = AitpAgent.generate();
  const manifest = agent.buildManifest({
    displayName: name,
    handshakeEndpoint: `http://localhost:${port}/aitp/handshake/`,
    offeredCaps: offers,
  });
  return { agent, manifest };
}

function fullHandshake(initiator, responder, respManifest, requested) {
  const sess = initiator.newSession();
  const rsess = responder.newResponder();
  const hello = sess.buildHello(respManifest, requested);
  const { ackJson: helloAck, sessionId } = rsess.processHello(hello);
  const commit = sess.processHelloAck(helloAck, sessionId);
  const { ackJson: commitAck } = rsess.processCommit(commit);
  // { tct, claims, grantVoucher } — the voucher is the delegation root.
  return sess.complete(commitAck);
}

test('delegation round-trip: A → B → C → A re-issues to C', () => {
  // Use ephemeral ports for clarity; manifests are not actually fetched.
  const { agent: a, manifest: aManifest } = buildPeer('A', 8101, ['demo.write']);
  const { agent: b } = buildPeer('B', 8102, ['demo.echo']);
  const { agent: c } = buildPeer('C', 8103, ['demo.read']);

  // B initiates against A and ends up holding A's TCT + grant voucher.
  const bCompleted = fullHandshake(b, a, aManifest, ['demo.write']);
  assert.ok(bCompleted.grantVoucher, 'A should issue a delegable grant voucher');

  // B mints a delegation from its voucher, binding C's AID, scoped to
  // demo.write. C's key binding is derived from c.aid itself.
  const delegationToken = b.buildDelegation(
    bCompleted.grantVoucher,
    c.aid,
    ['demo.write'],
  );

  // A verifies and mints a fresh TCT bound to C.
  const verified = verifyDelegation(delegationToken, a.aid);
  assert.equal(verified.delegator, a.aid);
  assert.equal(verified.delegatee, c.aid);
  assert.equal(verified.issuedBy, b.aid);
  assert.deepEqual(verified.grants, ['demo.write']);

  const freshTctForC = a.issueTctForDelegatee(verified);

  // C verifies the new TCT under the presented-TCT model (audience = C.aid).
  const ident = c.verifyTct(freshTctForC, 'demo.write', c.aid);
  assert.equal(ident.peerAid, a.aid);
  assert.ok(ident.grants.includes('demo.write'));
});

test('verifyDelegation rejects a wrong verifier AID', () => {
  const { agent: a, manifest: aManifest } = buildPeer('A', 8201, ['demo.write']);
  const { agent: b } = buildPeer('B', 8202, ['demo.echo']);
  const { agent: c } = buildPeer('C', 8203, ['demo.read']);
  const { agent: other } = buildPeer('Other', 8204, ['demo.x']);

  const bCompleted = fullHandshake(b, a, aManifest, ['demo.write']);
  const delegationToken = b.buildDelegation(
    bCompleted.grantVoucher,
    c.aid,
    ['demo.write'],
  );

  // A different agent's AID should be rejected as the verifier.
  assert.throws(() => verifyDelegation(delegationToken, other.aid));
});

test('verifyDelegation (strict default) rejects a multi-hop chain', () => {
  // RFC-AITP-0006 §4.4: any non-empty chain is rejected with
  // DELEGATION_MULTIHOP_NOT_SUPPORTED before any per-hop work.
  const { a, delegationToken } = buildVoucher();
  const tampered = injectMultihopChain(delegationToken);
  assert.throws(
    () => verifyDelegation(tampered, a.aid),
    /multi-hop delegation is not supported/,
  );
});

test(
  'verifyDelegationMultihop opts past the hop gate',
  { skip: !HAS_MULTIHOP ? 'built without multihop-delegation' : false },
  () => {
    // The opt-in verifier must get PAST the hop gate the strict path rejects
    // at — proven by failing with a *different* error (structure/signature)
    // rather than MULTIHOP_NOT_SUPPORTED.
    const { a, delegationToken } = buildVoucher();
    const tampered = injectMultihopChain(delegationToken);
    assert.throws(() => verifyDelegationMultihop(tampered, a.aid, 3), (err) => {
      assert.doesNotMatch(String(err.message), /multi-hop delegation is not supported/);
      return true;
    });
  },
);

// ── Revocation (RFC-AITP-0006 §4 step 7, RFC-AITP-0011 §6) ───────────────
//
// The Rust core has always had `revocation_check` / `hop_revocation_check`.
// What was missing was any way to reach them from JS: both entry points
// built their context with `VerifyDelegationContext::new`, which hardcodes
// both hooks to `None`, and neither exposed a parameter. Every delegation
// redeemed through this SDK therefore skipped step 7 — a MUST-reject — and a
// revoked source TCT still bought a freshly minted TCT for the delegatee.
//
// These pin the wiring; the semantics are covered in the core's own tests.

/** The `src_jti` of the voucher a delegation is rooted in — what A revokes. */
function srcJti(delegationToken) {
  const claims = decodeJwsPayload(delegationToken);
  return decodeJwsPayload(claims.voucher).src_jti;
}

test('revoking the source TCT rejects the delegation', () => {
  const { a, delegationToken } = buildVoucher();

  // Sanity: verifies while nothing is revoked, so the rejection below is
  // attributable to the deny list and nothing else.
  verifyDelegation(delegationToken, a.aid);

  assert.throws(
    () => verifyDelegation(delegationToken, a.aid, [srcJti(delegationToken)]),
    /revoked/i,
  );
});

test('a revoked source TCT yields no TCT for the delegatee', () => {
  // Step 7 is not bookkeeping: redeeming a delegation mints a *fresh TCT*
  // for the delegatee. Skipping it meant a revoked grant kept minting
  // credentials for a third party the grantor never re-authorized.
  const { a, delegationToken } = buildVoucher();

  assert.throws(() => {
    const verified = verifyDelegation(delegationToken, a.aid, [
      srcJti(delegationToken),
    ]);
    // Unreachable while step 7 holds. If it ever is reached, this is the
    // line that shows what the omission actually costs.
    a.issueTctForDelegatee(verified);
  });
});

test('an unrelated revoked jti does not reject', () => {
  // The lookup is keyed on `src_jti`, not "the list is non-empty" — without
  // this, a blanket-reject bug would pass the test above.
  const { a, delegationToken } = buildVoucher();
  const unrelated = '00000000-0000-4000-8000-000000000000';
  assert.notEqual(unrelated, srcJti(delegationToken));

  const verified = verifyDelegation(delegationToken, a.aid, [unrelated]);
  assert.equal(verified.delegator, a.aid);
});

test('omitting revokedJtis verifies a revoked delegation', () => {
  // Pins the default so it cannot change silently. Omitting the list waives
  // step 7 entirely — the same posture `verifyTct` takes, and the reason
  // both doc comments say verifiers SHOULD supply it. This documents the
  // hazard rather than endorsing it.
  const { a, delegationToken } = buildVoucher();
  const verified = verifyDelegation(delegationToken, a.aid);
  assert.equal(verified.delegator, a.aid);
});

test('malformed jtis in the list are ignored, not treated as matches', () => {
  // `parseRevokedSet` drops anything that is not a UUID. A list of pure
  // garbage must therefore behave exactly like an empty one, rather than
  // rejecting every token that comes near it.
  const { a, delegationToken } = buildVoucher();
  const verified = verifyDelegation(delegationToken, a.aid, [
    'not-a-uuid',
    '',
  ]);
  assert.equal(verified.delegator, a.aid);
});

test('revocation is consulted only after the signature checks', () => {
  // RFC-AITP-0008 §3.3 ordering, and it is not cosmetic: running the one
  // stateful lookup before the signature checks would let an unauthenticated
  // caller probe which jtis are in a verifier's deny list by watching which
  // forged tokens come back "revoked" versus "bad signature".
  const { a, delegationToken } = buildVoucher();
  const tampered = tamperJwsSignature(delegationToken);

  assert.throws(
    () => verifyDelegation(tampered, a.aid, [srcJti(delegationToken)]),
    (err) => !/revoked/i.test(String(err.message)),
    'a token failing signature verification was reported as revoked — the ' +
      'deny-list lookup ran before the signature checks',
  );
});
